lib/Decl.cpp

```diff
@@ -188,4 +195,131 @@ llvm::Value *FunctionDecl::CallFromLiftedBlock(
   }
 }
 
+// Serialize a FunctionDecl to JSON
+llvm::json::Object FunctionDecl::SerializeToJSON() {
+  llvm::json::Object json;
+
+  json.insert(
+      llvm::json::Object::KV{llvm::json::ObjectKey("name"), this->name});
+
+  if (FLAGS_demangle_names) {
+    json.insert(llvm::json::Object::KV{llvm::json::ObjectKey("demangled_name"),
+                                       this->demangled_name});
+  }
+
+  llvm::json::Array params_json;
+  for (auto pdecl : this->params) {
+    llvm::json::Value v = llvm::json::Value(pdecl.SerializeToJSON(*this->dl));
+    params_json.push_back(v);
+  }
+  json.insert(
+      llvm::json::Object::KV{llvm::json::ObjectKey("parameters"),
+                             llvm::json::Value(std::move(params_json))});
+
+  llvm::json::Array returns_json;
+  for (auto rdecl : this->returns) {
+    returns_json.push_back(llvm::json::Value(rdecl.SerializeToJSON(*this->dl)));
+  }
+  json.insert(
+      llvm::json::Object::KV{llvm::json::ObjectKey("return values"),
+                             llvm::json::Value(std::move(returns_json))});
+
+  if (this->return_stack_pointer) {
+    llvm::json::Object return_stack_pointer_json;
+    return_stack_pointer_json.insert(llvm::json::Object::KV{
+        llvm::json::ObjectKey("register"), this->return_stack_pointer->name});
+    return_stack_pointer_json.insert(llvm::json::Object::KV{
+        llvm::json::ObjectKey("offset"), this->return_stack_pointer_offset});
+    return_stack_pointer_json.insert(llvm::json::Object::KV{
+        llvm::json::ObjectKey("type"),
+        TranslateType(*this->return_stack_pointer->type, *this->dl)});
+
+    json.insert(llvm::json::Object::KV{
+        llvm::json::ObjectKey("return stack pointer"),
+        llvm::json::Value(std::move(return_stack_pointer_json))});
+  }
+  return json;
+}
+
+// Serialize a ParameterDecl to JSON
+llvm::json::Object ParameterDecl::SerializeToJSON(const llvm::DataLayout &dl) {
+  // Get the serialization for the ValueDecl
+  ValueDecl *val_decl_ptr = dynamic_cast<ValueDecl *>(this);
+  llvm::json::Object param_json = val_decl_ptr->SerializeToJSON(dl);
```

Comment:
What you want is `this->ValueDecl::SerializeToJSON(dl)` in place of teh `dynamic_cast`.